contrib_bots/lib/gcalendar.py

```diff
@@ -0,0 +1,254 @@
+import dateparser
+
+SCOPES = 'https://www.googleapis.com/auth/calendar'
+CLIENT_SECRET_FILE = '.gcal_client_secret.json'
+APPLICATION_NAME = 'Zulip Google Calendar Bot'
+
+credentials = None
+http = None
+service = None
+
+# NOTE ABOUT GOOGLE'S OAUTH
+# Right now there are some duplicates of this function in Zulip.
+#
+# For example, in this case it asks for user's consent to read and write
+# their calendars. Different bots/integrations that require other
+# Google Account permissions, have to repeat this process and create
+# new credentials files, which potentially fills the user's computer with
+# a lot files, spread all over the disk.
+#
+# The proper apporach for this would be to create a script that wraps
+# this procedure, and stores all the credentials in the same place.
+#
+# The ideal situation will be when Zulip becomes capable of managing API
+# keys and user credentials, allowing the user to choose between which
+# bots have access to what. However, this belongs more to the long-term
+# plan in Zulip's roadmap.
+
+def get_credentials():
+    '''Gets valid user credentials from storage.
+
+    If nothing has been stored, or if the stored credentials are invalid,
+    the OAuth2 flow is completed to obtain the new credentials.
+
+    Returns:
+        Credentials, the obtained credential.
+    '''
+    home_dir = os.path.expanduser('~')
+    credential_dir = os.path.join(home_dir, '.gcalbot-credentials')
+    if not os.path.exists(credential_dir):
+        os.makedirs(credential_dir)
+    credential_path = os.path.join(credential_dir,
+                                   'gcalbot.json')
+
+    store = Storage(credential_path)
+    cred = store.get()
+    if not cred or cred.invalid:
+        flow = client.flow_from_clientsecrets(os.path.join(home_dir, CLIENT_SECRET_FILE), SCOPES)
+        flow.user_agent = APPLICATION_NAME
+        flags = argparser.parse_args('--noauth_local_webserver --logging_level INFO'.split())
+        cred = tools.run_flow(flow, store, flags)
+    return cred
+
+def reply_message(msg, original_msg, client):
+    '''Replies to a specific message, through the channel of origin.
+
+    Keyword arguments:
+    msg -- content of the reply
+    original_msg -- message that is being replied
+    client -- client from the message handler
+    '''
+    message_type = original_msg['type']
+
+    if message_type == 'private':
+        original_sender = original_msg['sender_email']
+
+        client.send_message(dict(
+            type=message_type,
+            to=original_sender,
+            content=msg,
+        ))
+    elif message_type == 'stream':
+        original_stream = original_msg['display_recipient']
+        original_subject = original_msg['subject']
+
+        client.send_message(dict(
+            type=message_type,
+            to=original_stream,
+            subject=original_subject,
+            content=msg,
+        ))
+    else:
+        raise('Unknown message type')
```

Comment:
This will probably give error. I think you need to use something like `raise Exception('Unknown message type')`. 

Optional but better way would be to create your own exception class:
```
class UnknownMessageTypeError(Exception):
    pass

# ...

raise UnkownMessageTypeError
```